### DISCUSSION TITLE
Feature suggestion: automated rate limiting handling

### DISCUSSION THREAD

**Discussion by SachaG** [plhery/node-twitter-api-v2#86] [Q&A] [answered]
The library already does a pretty good job of making rate limits more explicit, but I was wondering if it could go one step further and actually delay your calls (for example with a new `fetchLastWhileWaiting()` function) to make them respect the rate limit? Or would that be beyond the scope of the library?

**Comment by alkihis** [comment]
Hi,
You can still implement by yourself a loop of `fetchLast` using [this model](https://github.com/PLhery/node-twitter-api-v2/blob/master/doc/rate-limiting.md#handle-errors---everywhere-in-this-library).

I don't personally think it's a good idea to allow a `Promise` to lock during multiple minutes... but it can be discussed :)

**Comment by SachaG** [comment]
Thanks. What kind of strategy would you use if you had to potentially run a script for multiple days because of the rate limits then? A cronjob wouldn't be able to adjust to always respect the shortest waiting time possible so it wouldn't be very efficient, but I guess it would be the simplest solution…

**Comment by alkihis** [comment, marked as answer]
You can await when there is a rate limit error, as suggested:

```ts
function sleep(ms: number) {
  return new Promise(resolve => setTimeout(ms, resolve));
}

async function autoRetryOnRateLimitError<T>(callback: () => T | Promise<T>) {
  while (true) {
    try {
      return await callback();
    } catch (error) {
      if (error instanceof ApiResponseError && error.rateLimitError && error.rateLimit) {
        const resetTimeout = error.rateLimit.reset * 1000; // convert to ms time instead of seconds time
        const timeToWait = resetTimeout - Date.now();

        await sleep(timeToWait);
        continue;
      }

      throw error;
    }
  }
}

// Get your paginator here
const paginator = await client.v2.following('12', { asPaginator: true, max_results: 1000 });
let currentPage = paginator;

while (true) {
  // Handle results of current page inside {currentPage.users}
  // Do stuff here...
  if (currentPage.users.length === 0) {
    break; // There's nothing, paginator ends :)
  }
 
  // Fetch the next page, and await if's there is a rate limit error
  currentPage = await autoRetryOnRateLimitError(() => paginator.next());
}
```

**Comment by SachaG** [comment]
Thanks for your help, I'll give this a try :)
